Realm/RLMDictionary.h

```diff
@@ -0,0 +1,416 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2021 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import <Realm/RLMCollection.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class RLMObject, RLMResults<RLMObjectType>, RLMDictionaryChange;
+
+/// A protocol defining a key type for RLMDictionary.
+@protocol RLMDictionaryKey <NSCopying>
```

Comment:
This isn't really how obj-c protocols work and doing this just makes the obj-c API a lot more obnoxious to call from swift without providing any additional safety.

Reply:
Are you talking about the protocol as a whole or the `NSCopying` part?

Reply:
The whole protocol.

Reply:
The reason why maps of objects have to be optional doesn't seem to apply to embedded objects, so it'd make sense to me to allow non-optional embedded objects but I assume that'll require some objectstore changes.